docs/spec/identities.md

```diff
@@ -0,0 +1,438 @@
+
+3. **Quorum**
+
+   The identity carrier passes 2., and is signed by a quorum of the keys
+   specified in the delegations of the document (`doc.delegations.len() / 2`).
+
+4. **Verified**
+
+    The identity carrier passes 3., and:
+
+    * The document does not refer to a previous revision, and no previous
+      revision is known
+    * **Or**, the set of signatures forms a quorum of the delegations of the
+      **previous** revision.
+
+    It is an error if:
+
+    * No previous revision is given, but a parent in the identity attestation
+      chain is found
+    * A previous revision is given, but the identity attestation chain does not
+      yield a parent
+    * A previous revision is given, but it is not the same the parent
+      attestation refers to
+    * The current and parent attestations refer to different `root`s
+
+The verification process can now be described recursively:
+
+```rust
+/* State transitions, definitions elided for brevity */
+fn signed(i: Untrusted<Identity>) -> Result<Signed<Identity>, Error>;
+fn quorum(i: Signed<Identity>) -> Result<Quorum<Identity>, Error>;
+fn verified(i: Quorum<Identity>, parent: Option<Verified<Identity>>)
+    -> Result<Verified<Identity>, Error>;
+
+fn verify(
+    head: Untrusted<Identity>,
+    mut parents: impl Iterator<Item = Untrusted<Identity>>,
+) -> Result<Verified<Identity>, Error> {
+    let head = quorum(signed(head)?)?;
+    let parent = match parents.next() {
+        None => Ok(None),
+        Some(parent) => verify(parent, parents),
+    }?;
+
+    verified(head, parent)
+}
+```
+
+> **Implementation Notes**:
+>
+> * The recusive definition is given for brevity, implementations may prefer to
+>   walk the history in reverse order.
+> * As an optimisation, implementations SHOULD store the result of a successful
+>   verification in persistent storage, such that verification of updates can
+>   start from the last-good state instead of the history root. Observe,
+>   however, that two different `Identity` objects (as identitified by their
+>   `id`) may attest the same `Doc` revision -- the persistent state SHOULD
+>   therefore allow retrieving the set of attesting `Identity`s by `Doc`
+>   revision.
+> * The `quorum` predicate above SHOULD actually skip over this parent instead
+>   of aborting the procedure -- it is possible that a later `Identity` reaches
+>   the quorum.
+
+## Delegations
+
+Radicle Link distinguishes two types of identities: personal and project. The
+first describes an actor in the system, while the second describes a (software)
+project on which one or more actors collaborate. Apart from their payload types
+`T`, they differ in their delegations type `D`:
+
+Personal identities can only delegate to anonymous keys, while project
+identities MAY attach a personal identity (by it's URN) to a key delegation.
+
+More formally:
+
+```rust
+type User<T> = Identity<T, HashSet<PublicKey>>;
+type Project<T> = Identity<T, HashMap<PublicKey, Option<Urn>>>;
```

Comment:
This can be controversial.

Linking keys instead of URNs has the advantage of making the delegated keys explicit, but it has the disadvantage of forcing project updates whenever a delegated user revokes a key.

It also has the disadvantage that if a user creates a new device key (and adds it to their key set in their identity) that device key cannot be used to emit events in a project feed, even if the user were a project maintainer (through a key listed in the project delegates set).

In principle a maintainer should be a _user_ and not a key.